mri/utils/process_SWI

```diff
@@ -0,0 +1,127 @@
+import matplotlib.pyplot as plt
+import numpy as np
+
+#functions
+def virtual_coil_reconstruction(imgs):
+    """
+    Calculate the combination of all the coils thanks to the virtual coil
+    methods as defined:
+    Parameters
+    ----------
+    imgs: np.ndarray
+        The images reconstructed channel by channel [Nch, Nx, Ny, Nz]
+    Returns
+    -------
+    I: np.ndarray
+        The combination of all the channels in a complex valued [Nx, Ny, Nz]
+    """
+    # Compute first the virtual coil
+    nch, nx, ny, nz = imgs.shape
+    weights = np.sum(np.abs(imgs), axis=0)
+    weights[weights==0] = 1e-16
+    phase_reference = np.asarray([np.angle(np.sum(
+        imgs[ch].flatten())) for ch in range(nch)])
+    reference = np.asarray([(imgs[ch] / weights) / np.exp(1j * phase_reference[ch])
+                           for ch in range(32)])
+    virtual_coil = np.sum(reference, axis=0)
+    difference_original_vs_virtual = np.asarray(
+        [np.conjugate(imgs[ch]) * virtual_coil for ch in range(nch)])
+    hanning_1d = np.expand_dims(np.hanning(np.minimum(nx,ny)), 1)
+    hanning_2d = np.fft.fftshift(np.dot(hanning_1d, hanning_1d.T))
+    hanning_Nd = np.tile(np.expand_dims(np.tile(hanning_2d, (nch, 1, 1)), -1), (1, 1, 1, nz))
+    # Removing the background noise via low pass filtering
+    difference_original_vs_virtual = np.fft.ifft2(np.fft.fft2(
+        difference_original_vs_virtual, axes=(1, 2)) * hanning_Nd, axes=(1, 2))
+    I = np.asarray([imgs[ch] * np.exp(1j * np.angle(
+                    difference_original_vs_virtual[ch])) for ch in range(nch)])
+    return np.sum(I, 0)
+
+def get_hanning_filter(volume_shape):
+    """
+    Parameters
+    ----------
+    volume_shape: np.ndarray
+        size of the resulting hanning filter
+    Returns
+    -------
+    imRef_homodyne: np.ndarray
+        the hanning filter matrix
+    """
+
+    nx, ny, nz = volume_shape
+
+    # create hanning filter 3D
+    hanning_1d = np.zeros([np.minimum(nx,ny),1])
+    hanning_1d[int(np.round(3 * nx / 10)):int(np.round(7 * nx / 10))] = \
+        np.expand_dims(np.hanning(int(np.round(7 * nx / 10))-int(np.round(3 * nx / 10))), 1)
+    hanning_2d = np.fft.fftshift(np.dot(hanning_1d, hanning_1d.T))
+    hanning_Nd = np.tile(np.expand_dims(hanning_2d, -1), (1,1,nz))
+
+    return hanning_Nd
+
+def get_SWI(imRef,  verbose=100) :
+    """
+        Parameters
+        ----------
+        image_ch: np.ndarray
+            Images reconstructed per channels in shape [Nch, Nx, Ny, Nz] in 3D or [Nch, Nx, Ny] in 2D
+        Returns
+        -------
+        SWI: np.ndarray
+            SWI image
+    """
+    if len(imRef.shape) == 3 :
+        image=imRef[:,:,:,np.newaxis]
```

Comment:
I'm not very comfortable with this, let us see if we can get hanning filter generateable for an arbitrary dimension using loop? 
We could resort to this as a last resort.